include/sound/sof/control.h

```diff
@@ -149,7 +149,9 @@ struct sof_ipc_comp_event {
 	/* control data - add new types if needed */
 	union {
 		/* data can be used by binary controls */
-		struct sof_abi_hdr data[0];
+		DECLARE_FLEX_ARRAY(u8, data); /* flexible array of
```

Comment:
I am not sure I understand how this would work with all the sizeof that depend on 'struct sof_abi_hdr', is this completely independent/safe?

````
ipc3-control.c: /* scontrol->max_size has been verified to be >= sizeof(struct sof_abi_hdr) */
ipc3-control.c: if (cdata->data->size > scontrol->max_size - sizeof(struct sof_abi_hdr)) {
ipc3-control.c:                             scontrol->max_size - sizeof(struct sof_abi_hdr));
ipc3-control.c: data_size = cdata->data->size + sizeof(struct sof_abi_hdr);
ipc3-control.c: /* be->max has been verified to be >= sizeof(struct sof_abi_hdr) */
ipc3-control.c: if (cdata->data->size > scontrol->max_size - sizeof(struct sof_abi_hdr)) {
ipc3-control.c: if (cdata->data->size > scontrol->max_size - sizeof(struct sof_abi_hdr)) {
ipc3-control.c:                             scontrol->max_size - sizeof(struct sof_abi_hdr));
ipc3-control.c: data_size = cdata->data->size + sizeof(struct sof_abi_hdr);
ipc3-control.c:         expected_size += cdata->num_elems + sizeof(struct sof_abi_hdr);
ipc3-topology.c:        if (scontrol->max_size < (sizeof(*cdata) + sizeof(struct sof_abi_hdr))) {
ipc3-topology.c:                if (cdata->data->size + sizeof(struct sof_abi_hdr) != scontrol->priv_size) {
ipc3.c:         hdr_bytes = sizeof(struct sof_ipc_ctrl_data) + sizeof(struct sof_abi_hdr);
````

Reply:
The issue is that no matter how you wrap the sof_abi_hdr struct, it is always going to generate warning of flexible array as within the sof_abi_hdr we have at the end `data[]`.
That makes it flexible in itself and if you try to create a flexible array of a flexible struct, it is never going to make any sense and fundamentally wrong...
There is no such a thing as:
```c
struct sof_abi_hdr *hdr = comp->data;

if (hdr[1].size == 12)
...
```
hdr[1] is invalid as the second sof_abi_hdr in that area is at ((u8 *)comp->data) + sizeof(struct sof_abi_hdr) + hdr->size;

I have not found any other way than to convert it to u8 array and document that we need to use sof_abi_hdr ont he area to make sense of it.